🐛Source-mssql: added support for some datatypes

etsybaev opened the pull request:
## What
Currently, we have 2 different classes for integration datatype tests CdcMssqlSourceDatatypeTest and MssqlSourceDatatypeTest. They have the same input args, but differ "expected result" for some types.
Some of Cdc types were not supported at all.

## How
Aligned tests for normal and CDC datatype tests to make those tests ready for merging.
This PR is basically the first part of splitting this https://github.com/airbytehq/airbyte/pull/14379 into 2 PRs as it had been requested.  Once it appears to be merged, I will create a second PR for tests merging to have the same tests for both sync ways (to verify that output is the same for both). 


Normal sync:
- Aligned time format to what we have in other connectors.

CDC:
- Added proper handling: binary, varbinary, geometry, geography, 
- Fixed: time, datetimeoffset, 


## Recommended reading order
1. airbyte-integrations/bases/debezium-v1-4-2/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java


## 🚨 User Impact 🚨
Shouldn't introduce any breaking changes.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [x] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

etsybaev commented:
/test connector=connectors/source-mssql

grishick reviewed (changes requested):
Please specify which data types are added in this PR as part of the commit message and PR description.

I see that you removed trailing "0"s from one of the tests - why?

etsybaev commented:
@grishick the initial idea was to do not transform the user's data, i.e. if user had time with millis - then we would also add it, but if user's data was without millis - then also not add it. But I got your concern and updated PR to have millis by default (re-used pattern that we already use in other connectors). 

Also updated description as per your request. 
Thanks

etsybaev commented:
/test connector=connectors/source-mssql

gsmith-schlesinger commented:
Hey, so this is actually related to work I'm currently trying to do in a fork. 

I know your PR doesn't capture "MONEY", "NUMERIC", and "DECIMAL". But could we tack those changes in?

https://github.com/airbytehq/airbyte/issues/12949

gsmith-schlesinger commented on airbyte-integrations/bases/debezium-v1-4-2/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java:
```diff
@@ -34,11 +46,61 @@ public void converterFor(final RelationalColumn field,
       registerDate(field, registration);
     } else if (SMALLMONEY_TYPE.equalsIgnoreCase(field.typeName())) {
```
```suggestion
    } else if (DECIMAL_TYPE.contains(field.typeName().toUpperCase())) {
```

etsybaev commented on airbyte-integrations/bases/debezium-v1-4-2/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java:
```diff
@@ -20,9 +27,14 @@ public class MSSQLConverter implements CustomConverter<SchemaBuilder, Relational
+  private final Set<String> BINARY = Set.of("VARBINARY", "BINARY");
+  private static final String DATETIMEOFFSET = "DATETIMEOFFSET";
+  private static final String TIME_TYPE = "TIME";
+  private static final String SMALLMONEY_TYPE = "SMALLMONEY";
```
Hi @gsmith-schlesinger .
Could you please clarify what's wrong with those types?
If we update it a is then some dataType tests started failing.
Also if we update something for CDD migration we need to make sure we do the same for normal sync. In the nearest future we'll have the same tests for both Normal and CDC syncs. 


Ex.:
org.opentest4j.AssertionFailedError: Returned value '999.0' from stream dbo_1_decimal is not in the expected list: [999.00, 5.10, 0.00, null] ==> expected: <true> 


Thanks

gsmith-schlesinger replied:
So currently as mentioned in #12949 and from this issues #5609 from are exactly the same issue we're currently experiencing the difference is the column type from the database we're reading from is use a `MONEY`. 

There's a PR which refers to this area of code that "fixed" the issue for `SMALLMONEY` but it didn't encapsulate all _decimal_ type columns from MSSQL. 

You'll see the error logs in both tickets, and those error logs are the same thing we're seeing in ours. 

`Failed to convert JSON to Avro: Could not evaluate union, field CTR is expected to be one of these: NULL, DOUBLE. If this is a complex type, check if offending field (path: CTR) adheres to schema: 0.00`

There's a _workaround_ using CSV or JSON. But that's not sufficient for our use case.

gsmith-schlesinger replied:
Should note as well this issue comes up _only_ on CDC / Incremental airbyte tasks.  The Full Refresh / Overwrite task works successfully.

etsybaev replied:
do you have any PR where you've fixed and aligned it with regular sync?
Just asking as this particular PR basically is the first step of merging tests for normal and CDC sync's to make sure that behavior (output) are the same - https://github.com/airbytehq/airbyte/pull/14379.
I just hab need asked to split it into 2 different PRs.


Thanks

gsmith-schlesinger replied:
I do not yet, but I'll honestly have to wait for your PR to be merged before I can submit mine anyways as you've made a lot of changes I was already planning on making in my own fork. I only started working towards this two days ago.

subodh1810 reviewed (changes requested):
There are 2 versions of MSSQLConverter, can you replicate your changes to `airbyte-integrations/bases/debezium-v1-9-2/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java` as well. 

Also please take a look at the issue https://github.com/airbytehq/airbyte/issues/14628, we want to avoid mutating the data types in any way possible so please verify that its not happening in this PR

grishick commented on airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/CdcMssqlSourceDatatypeTest.java:
```diff
@@ -391,32 +391,27 @@ protected void initTests() {
             .airbyteType(JsonSchemaType.STRING)
             .addInsertValues("'0001-01-10 00:00:00 +01:00'", "'9999-01-10 00:00:00 +01:00'", "null")
-            .addExpectedValues("0001-01-10 00:00:00 +01:00", "9999-01-10 00:00:00 +01:00", null)
+            .addExpectedValues("0001-01-10 00:00:00.0000000 +01:00",
```
why do we expect to have milliseconds when insert values do not have them?

etsybaev replied:
I thought this was your concern in the comment above https://github.com/airbytehq/airbyte/pull/14449#pullrequestreview-1030820583
Previously we had millis in any case. I removed it for cases when source didn't have it. I thought you had a concern about it so I added it back, but aligned it to the format that we have for other connectors. And also aligned between normal and CDC cync. Because for now we have different behavior even inside mssql connector.  If it's not OK, kindly ask to specify your expected behavior. Thanks

grishick commented on airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/CdcMssqlSourceDatatypeTest.java:
```diff
@@ -449,17 +444,14 @@ protected void initTests() {
-            // "POLYGON ((0 0, 150 0, 150 150, 0 150, 0 0)", null)
+            .addInsertValues("geometry::STGeomFromText('LINESTRING (100 100, 20 180, 180 180)', 0)",
+                "null")
+            .addExpectedValues("LINESTRING(100 100, 20 180, 180 180)", null)
```
why do we expect `null` in place of `0` is this the correct behavior?

etsybaev replied:
![Selection_359](https://user-images.githubusercontent.com/4688354/178663921-05bc79b8-aa1c-412d-836b-659977fc8db9.png)
We put null and expect null. There are 2 test cases:
1. "geometry::STGeomFromText('LINESTRING (100 100, 20 180, 180 180)', 0)"
2. null

grishick commented on airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/MssqlSourceDatatypeTest.java:
```diff
@@ -225,7 +225,7 @@ protected void initTests() {
             .airbyteType(JsonSchemaType.STRING)
             .addInsertValues("null", "'13:00:01'", "'13:00:04Z'", "'13:00:04.123456Z'")
-            .addExpectedValues(null, "13:00:01.0000000", "13:00:04.0000000", "13:00:04.1234560")
+            .addExpectedValues(null, "13:00:01.000000", "13:00:04.000000", "13:00:04.123456")
```
same here - is it the correct behavior to expect milliseconds when inserting data w/o milliseconds?

etsybaev replied:
I thought this was your concern in the comment above https://github.com/airbytehq/airbyte/pull/14449#pullrequestreview-1030820583
Previously we had millis in any case. I removed it for cases when source didn't have it.  I thought you had a concern about it so I added it back, but aligned it to the format that we have for other connectors. If it's not OK, kindly ask to specify your expected behavior. Thanks

grishick replied:
This was a misunderstanding. I wasn't expressing a concern, but asking for the underlying logic/reason for removing trailing zeros in some of the test cases. Generally speaking, we should not be changing source data. If source has time w/o millis we should not be adding them, if source has time with millis, we should not be removing them.

grishick commented on airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/MssqlSourceDatatypeTest.java:
```diff
@@ -225,7 +225,7 @@ protected void initTests() {
             .sourceType("time")
             .airbyteType(JsonSchemaType.STRING)
             .addInsertValues("null", "'13:00:01'", "'13:00:04Z'", "'13:00:04.123456Z'")
```
It looks like we expect to lose "Z" - is this the correct behavior?

etsybaev replied:
I just aligned it to what we already use in other connectors (re-used existing time format). Should I add an additional format to add it back? Thanks

etsybaev commented:
> subodh1810

Hi @subodh1810 .
My changes were supposed to to also update it.
Basically, the initial idea was to alight it in normal and CDC sync + merge the tests, but I was asked to split it into 2 PRs (https://github.com/airbytehq/airbyte/pull/14379). For now we have separate tests for normal and CDC sync. But those sync pretty often may produce different outputs.
Are you going to also align it in your PR? I mean if we set some format for CDC sync - then also make sure that normal sync will produce the exactly same output?
Thanks

etsybaev commented:
Blocked by https://github.com/airbytehq/airbyte/issues/14628

duytp commented:
@gsmith-schlesinger, you save my days.
I got the same errors when replicating MSSQL to S3 (incremental | append).

Your sugesstion resovled the issue.

---------------------------------------------------------------------
Additional Failure Information: tech.allegro.schema.json2avro.converter.AvroConversionException: Failed to convert JSON to Avro: Could not evaluate union, field TotalAmount is expected to be one of these: NULL, DOUBLE. If this is a complex type, check if offending field (path: TotalAmount) adheres to schema: 0.00

A PR is recommended.
